File: algorithms/Maximum Product of Three Numbers/Maximum Product of Three Numbers.py
class Solution:
    def maximumProduct(self, nums: List[int]) -> int:
        # TC = O(NlogN) because sorting the array 
        # SC = O(1); no extra space needed; sorting was done in place.
        
        # sorting the array in descending order
        nums.sort(reverse = True)
        
        # maximum product can only occur for:
        # 1. positive no * positive no * positive no
        # 2. negative no * negative no * positive no
        
        # one negative and two positives and all negatives wont give max product
        # case where all numbers in the array are negative 
        # eg : [-4,-3,-2,-1] is covered in all positives 
        
        return max(nums[0]*nums[1]*nums[2],nums[-1]*nums[-2]*nums[0])

File: algorithms/M/Maximum Product of Three Numbers/Maximum Product of Three Numbers.js
var maximumProduct = function(nums) {
    nums.sort((a, b) => a-b)
    
    var lastNumber = nums.length - 1
    var midNumber = nums.length - 2
    var firstNumber = nums.length - 3
    var total = nums[lastNumber] * nums[midNumber] * nums[firstNumber]
    return total
    
};
